File: app/src/main/java/com/originb/inkwisenote2/modules/handwrittennotes/data/HandwrittenNoteEntity.java
package com.originb.inkwisenote2.modules.handwrittennotes.data;

import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.ForeignKey;
import androidx.room.PrimaryKey;
import com.originb.inkwisenote2.modules.smartnotes.data.AtomicNoteEntity;
import com.originb.inkwisenote2.modules.smartnotes.data.SmartBookEntity;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@NoArgsConstructor
@Entity(tableName = "handwritten_notes",
        foreignKeys = {
                @ForeignKey(entity = SmartBookEntity.class,
                        parentColumns = "book_id",
                        childColumns = "book_id",
                        onDelete = ForeignKey.CASCADE),
                @ForeignKey(entity = AtomicNoteEntity.class,
                        parentColumns = "note_id",
                        childColumns = "note_id",
                        onDelete = ForeignKey.CASCADE)
        })
public class HandwrittenNoteEntity {

    @PrimaryKey()
    @ColumnInfo(name = "note_id")
    private long noteId;

    @ColumnInfo(name = "book_id")
    private long bookId;

    @ColumnInfo(name = "bitmap_file_path")
    private String bitmapFilePath;

    @ColumnInfo(name = "bitmap_hash")
    private String bitmapHash;

    @ColumnInfo(name = "page_template_file_path")
    private String pageTemplateFilePath;

    @ColumnInfo(name = "page_template_hash")
    private String pageTemplateHash;

    @ColumnInfo(name = "created_time_ms")
    private long createdTimeMillis;

    @ColumnInfo(name = "last_modified_time_ms")
    private long lastModifiedTimeMillis;
}

File: app/src/main/java/com/originb/inkwisenote2/modules/handwrittennotes/data/HandwrittenNoteEntity.kt
package com.originb.inkwisenote2.modules.handwrittennotes.data

import androidx.room.*
import com.originb.inkwisenote2.modules.smartnotes.data.AtomicNoteEntity
import com.originb.inkwisenote2.modules.smartnotes.data.SmartBookEntity

@Entity(
    tableName = "handwritten_notes",
    foreignKeys = [ForeignKey(
        entity = SmartBookEntity::class,
        parentColumns = "book_id",
        childColumns = "book_id",
        onDelete = ForeignKey.Companion.CASCADE
    ), ForeignKey(
        entity = AtomicNoteEntity::class,
        parentColumns = "note_id",
        childColumns = "note_id",
        onDelete = ForeignKey.Companion.CASCADE
    )],
    indices = [Index("book_id"), Index("note_id")]
)
class HandwrittenNoteEntity  // No-args constructor required by Room
{
    // Getters and Setters
    @PrimaryKey
    @ColumnInfo(name = "note_id")
    var noteId: Long = 0

    @ColumnInfo(name = "book_id")
    var bookId: Long = 0

    @ColumnInfo(name = "bitmap_file_path")
    var bitmapFilePath: String? = null

    @ColumnInfo(name = "bitmap_hash")
    var bitmapHash: String? = null

    @ColumnInfo(name = "page_template_file_path")
    var pageTemplateFilePath: String? = null

    @ColumnInfo(name = "page_template_hash")
    var pageTemplateHash: String? = null

    @ColumnInfo(name = "created_time_ms")
    var createdTimeMillis: Long = 0

    @ColumnInfo(name = "last_modified_time_ms")
    var lastModifiedTimeMillis: Long = 0
} // todo: Add width and height of the drawingView
// then use logical space to create the canvas
// check notion for details

